Review followup: PR #26263 — fix: pause repeated worker recovery loops

## Unaddressed review bot suggestions

PR #26263 was merged with unaddressed review bot feedback. Each comment
below includes its file path, line number, a direct link to the inline
review comment, and a `diff` fence with the code context the bot was
flagging. Resolved and outdated threads are filtered out via GitHub's
GraphQL review-thread state. Read the relevant lines, decide whether
the suggestion is correct, and either apply the fix or close this issue
with a wontfix rationale.

**Source PR:** https://github.com/marcusquinn/aidevops/pull/26263

---

### You are the triager (worker-is-triager rule)

This issue is **auto-created from review bot output** and dispatched
directly to you. Review bots can be wrong: hallucinated line refs, false
premises about codebase structure, template-driven sweeps without
measurements (see GH#17832-17835 for prior art and `AGENTS.md`
"AI-Generated Issue Quality"). **Do not assume the bot is correct.** Verify before acting.

You must end in exactly one of three outcomes — no fourth "hand it back
to the human" path exists. Humans approve decisions; they do not re-do
analysis.

#### Outcome A — Premise falsified → close the issue

1. Read the cited `file:line` (listed under *Files to modify* below).
2. If the bot's claim is factually wrong (file doesn't exist at that
   line, function doesn't behave as described, "auto-generated" section
   isn't actually auto-generated, etc.), **close the issue** with a
   comment in this shape:

   > **Premise falsified.** \<what the bot claimed\>. \<what the code
   > actually shows, with a `file:line` citation or one-line quote\>.
   > Not acting.

   No PR. No further dispatch. The closing comment trains the next
   session reading this thread and the noise filter.

#### Outcome B — Premise correct + fix is obvious → implement and PR

1. Verify the bot's premise as above.
2. Read the Worker Guidance section below, open a worktree, implement.
3. Open a PR with `Resolves #<this-issue-number>` in the body
   (use THIS issue's number, not the source PR's) so merge auto-closes it.
4. Follow the normal Lifecycle Gate (brief, tests, review-bot-gate,
   merge, postflight).

#### Outcome C — Premise correct but approach is a genuine judgment call

Only use this path if you reach it after Outcomes A and B don't apply:
the bot's finding is real, but the fix requires a decision that is
architectural, policy, breaking-change, or otherwise genuinely outside
what you can resolve autonomously. In that case, post a **decision
comment** with exactly these fields:

- **Premise check:** one line, confirming the finding is real.
- **Analysis:** 2-4 bullets on the trade-offs.
- **Recommended path:** the option you would take if the decision were
  yours, with rationale.
- **Specific question:** the single decision the human needs to make
  (yes/no or pick-one, not open-ended).

Then apply `needs-maintainer-review` and stop. The human wakes up to a
ready-to-approve recommendation, not a blank task.

> **Ambiguity about scope or style is not Outcome C.** Per
> `AGENTS.md` "Reasoning responsibility", the model does the
> thinking and delivers a recommendation. Only escalate what is genuinely
> a maintainer-only decision.

### Worker Guidance

**Files to modify:**

- `.agents/scripts/dispatch-dedup-recovery-loop.sh:109`

**Implementation steps (Outcome B path):**

1. Read the `diff` block under each inline comment below — it shows the
   exact code the bot was flagging. Open the file only if you need
   surrounding context beyond what the diff tail shows.
2. Read the bot's full comment below the diff — it contains the rationale
   and any suggested change.
3. Verify the premise before implementing (see Outcome A). If the premise
   is wrong, switch to Outcome A instead of burning iterations trying to
   satisfy a wrong suggestion.
4. If multiple comments target the same file, group your edits into one
   logical commit.
5. Run `shellcheck` / `markdownlint-cli2` / project tests as appropriate.

**Verification:**

- Open the new PR with `Resolves #<this-issue>` so this followup is auto-closed on merge.
- If the bot's suggestion was incorrect, close this issue with a Outcome A comment — do not open a no-op PR.

### Inline comments

#### gemini-code-assist on `.agents/scripts/dispatch-dedup-recovery-loop.sh:109`

[View inline comment](https://github.com/marcusquinn/aidevops/pull/26263#discussion_r3510234978)

```diff
+	local existing_block="0"
+
+	set_issue_status "$issue_number" "$repo_slug" "" --add-label "needs-maintainer-review" >/dev/null 2>&1 || true
+
+	existing_block=$(_ddh_count_recovery_loop_blocks "$comments_json")
+	if [[ "$existing_block" -eq 0 ]]; then
+		local diag=""
+		# shellcheck disable=SC2016 # Backticks are literal Markdown, not command substitution.
+		diag=$(printf '<!-- ops:start — workers: skip this comment, it is audit trail not implementation context -->\n<!-- worker-recovery-loop:blocked count=%s threshold=%s window_s=%s latest=%s -->\n<!-- dispatch-circuit-breaker:worker_recovery_loop -->\n## Dispatch paused: repeated worker recovery failures\n\nThis issue has produced %s worker recovery-failure outcome(s) within the last %s seconds. Those outcomes mean workers produced branch evidence but the automation could not confirm a PR, so another redispatch would likely add more audit comments without solving the issue.\n\n**Action:** applied `needs-maintainer-review` and cleared active status labels. Automated dispatch is suspended until a human reviews the runner worktrees/logs, recovers any branch output, or lands a setup-side fix.\n\n**Verification before re-enabling:** confirm a PR exists for the work or confirm there is no unrecovered worker branch output left to preserve, then remove `needs-maintainer-review`.\n<!-- ops:end -->' \
+			"$count" "$threshold" "$window_s" "${latest_iso:-unknown}" "$count" "$window_s")
+		gh api "$comments_post_endpoint" --method POST --field body="$diag" >/dev/null 2>&1 || true
+	fi
```

> ![high](https://www.gstatic.com/codereviewagent/high-priority.svg)
> 
> The diagnostic comment posted by `_ddh_apply_recovery_loop_hold` does not include the `aidevops` version. This prevents `_nmr_breaker_release_retry_reason` in `pulse-nmr-approval.sh` from extracting the version and allowing an automated retry after a newer `aidevops` release is deployed, which contradicts the stated PR goal of "allowing retry after a newer aidevops release".
> 
> Please retrieve the current `aidevops` version and include it in the HTML comment metadata (e.g., `version=%s`). Avoid redundant pre-checks before calling `aidevops_find_version` since the call is non-fatal and can be delegated directly.
> 
> ```suggestion
> 	if [[ "$existing_block" -eq 0 ]]; then
> 		local version=""
> 		version=$(aidevops_find_version 2>/dev/null || true)
> 		if [[ -z "$version" && -f "${AGENTS_DIR:-$HOME/.aidevops/agents}/VERSION" ]]; then
> 			version=$(tr -d '[:space:]' <"${AGENTS_DIR:-$HOME/.aidevops/agents}/VERSION" 2>/dev/null || true)
> 		fi
> 		local diag=""
> 		# shellcheck disable=SC2016 # Backticks are literal Markdown, not command substitution.
> 		diag=$(printf '<!-- ops:start — workers: skip this comment, it is audit trail not implementation context -->\n<!-- worker-recovery-loop:blocked count=%s threshold=%s window_s=%s latest=%s version=%s -->\n<!-- dispatch-circuit-breaker:worker_recovery_loop -->\n## Dispatch paused: repeated worker recovery failures\n\nThis issue has produced %s worker recovery-failure outcome(s) within the last %s seconds. Those outcomes mean workers produced branch evidence but the automation could not confirm a PR, so another redispatch would likely add more audit comments without solving the issue.\n\n**Action:** applied needs-maintainer-review and cleared active status labels. Automated dispatch is suspended until a human reviews the runner worktrees/logs, recovers any branch output, or lands a setup-side fix.\n\n**Verification before re-enabling:** confirm a PR exists for the work or confi

### PR review summaries

#### gemini-code-assist[bot] review summary

[View review](https://github.com/marcusquinn/aidevops/pull/26263#pullrequestreview-4614259453)

> ## Code Review
> 
> This pull request introduces an issue-level worker recovery loop fuse to pause automated dispatch upon repeated worker recovery failures across branches, and refactors the daily quality sweep to upsert a rolling summary comment instead of appending new ones. The review feedback correctly identifies that the recovery loop hold diagnostic comment should include the `aidevops` version to enable automated retries after a newer release is deployed.
> 
> > [!IMPORTANT]
> > The [consumer version of Gemini Code Assist on GitHub](https://developers.google.com/gemini-code-assist/docs/review-repo-code) is being sunset. Starting **June 18, 2026**, new organization installations will be blocked, and all code review activity will officially cease on **July 17, 2026**.
> > For more details on the timeline and next steps, please review the [Help Documentation](https://developers.google.com/gemini-code-assist/docs/deprecations/consumer-code-review).
<!-- aidevops:sig -->
---
[aidevops.sh](https://aidevops.sh) v3.31.19 automated scan.

--- a/.agents/scripts/dispatch-dedup-recovery-loop.sh
+++ b/.agents/scripts/dispatch-dedup-recovery-loop.sh
@@ -3,6 +3,12 @@
 # SPDX-FileCopyrightText: 2025-2026 Marcus Quinn
 # dispatch-dedup-recovery-loop.sh — issue-level worker recovery loop fuse.
 
+_DDH_RECOVERY_LOOP_SCRIPT_DIR="$(cd "$(dirname "${BASH_SOURCE[0]:-$0}")" 2>/dev/null && pwd)" || _DDH_RECOVERY_LOOP_SCRIPT_DIR=""
+if [[ -n "$_DDH_RECOVERY_LOOP_SCRIPT_DIR" && -r "${_DDH_RECOVERY_LOOP_SCRIPT_DIR}/lib/version.sh" ]]; then
+	# shellcheck source=lib/version.sh
+	source "${_DDH_RECOVERY_LOOP_SCRIPT_DIR}/lib/version.sh"
+fi
+
 #######################################
 # Count recent worker recovery failures across branches.
 #
@@ -101,10 +107,16 @@ _ddh_apply_recovery_loop_hold() {
 
 	existing_block=$(_ddh_count_recovery_loop_blocks "$comments_json")
 	if [[ "$existing_block" -eq 0 ]]; then
+		local version=""
+		version=$(aidevops_find_version 2>/dev/null || true)
+		if [[ -z "$version" && -r "${AGENTS_DIR:-$HOME/.aidevops/agents}/VERSION" ]]; then
+			version=$(tr -d '[:space:]' <"${AGENTS_DIR:-$HOME/.aidevops/agents}/VERSION" 2>/dev/null || true)
+		fi
+		version="${version:-unknown}"
 		local diag=""
 		# shellcheck disable=SC2016 # Backticks are literal Markdown, not command substitution.
-		diag=$(printf '<!-- ops:start — workers: skip this comment, it is audit trail not implementation context -->\n<!-- worker-recovery-loop:blocked count=%s threshold=%s window_s=%s latest=%s -->\n<!-- dispatch-circuit-breaker:worker_recovery_loop -->\n## Dispatch paused: repeated worker recovery failures\n\nThis issue has produced %s worker recovery-failure outcome(s) within the last %s seconds. Those outcomes mean workers produced branch evidence but the automation could not confirm a PR, so another redispatch would likely add more audit comments without solving the issue.\n\n**Action:** applied `needs-maintainer-review` and cleared active status labels. Automated dispatch is suspended until a human reviews the runner worktrees/logs, recovers any branch output, or lands a setup-side fix.\n\n**Verification before re-enabling:** confirm a PR exists for the work or confirm there is no unrecovered worker branch output left to preserve, then remove `needs-maintainer-review`.\n<!-- ops:end -->' \
-			"$count" "$threshold" "$window_s" "${latest_iso:-unknown}" "$count" "$window_s")
+		diag=$(printf '<!-- ops:start — workers: skip this comment, it is audit trail not implementation context -->\n<!-- worker-recovery-loop:blocked count=%s threshold=%s window_s=%s latest=%s version=%s -->\n<!-- dispatch-circuit-breaker:worker_recovery_loop -->\n## Dispatch paused: repeated worker recovery failures\n\nThis issue has produced %s worker recovery-failure outcome(s) within the last %s seconds. Those outcomes mean workers produced branch evidence but the automation could not confirm a PR, so another redispatch would likely add more audit comments without solving the issue.\n\n**Action:** applied `needs-maintainer-review` and cleared active status labels. Automated dispatch is suspended until a human reviews the runner worktrees/logs, recovers any branch output, or lands a setup-side fix.\n\n**Verification before re-enabling:** confirm a PR exists for the work or confirm there is no unrecovered worker branch output left to preserve, then remove `needs-maintainer-review`.\n<!-- ops:end -->' \
+			"$count" "$threshold" "$window_s" "${latest_iso:-unknown}" "$version" "$count" "$window_s")
 		gh api "$comments_post_endpoint" --method POST --field body="$diag" >/dev/null 2>&1 || true
 	fi
 

--- a/.agents/scripts/tests/test-worker-recovery-loop.sh
+++ b/.agents/scripts/tests/test-worker-recovery-loop.sh
@@ -120,11 +120,12 @@ GHEOF
 test_cross_branch_recovery_failures_block() {
 	local output=""
 	if output=$("$HELPER_SCRIPT" check-recovery-loop 100 owner/repo 2>/dev/null); then
-		if [[ "$output" == *"WORKER_RECOVERY_LOOP_BLOCKED"* && -f "${TEST_ROOT}/posts/100.argv" && -f "${TEST_ROOT}/edits/100.argv" ]]; then
+		if [[ "$output" == *"WORKER_RECOVERY_LOOP_BLOCKED"* && -f "${TEST_ROOT}/posts/100.argv" ]] &&
+			grep -q 'worker-recovery-loop:blocked .* version=[0-9]' "${TEST_ROOT}/posts/100.argv"; then
 			print_result "cross-branch recovery failures trip issue-level hold" 0
 			return 0
 		fi
-		print_result "cross-branch recovery failures trip issue-level hold" 1 "Unexpected output or missing side effects: ${output}"
+		print_result "cross-branch recovery failures trip issue-level hold" 1 "Unexpected output, missing diagnostic post, or missing version metadata: ${output}"
 		return 0
 	fi
 	print_result "cross-branch recovery failures trip issue-level hold" 1 "Expected dispatch hold"

